File: 2025-03-29/Course_Schedule.java
package amazon;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;

public class Course_Schedule {
    public static boolean dfs(List<Integer>[] adj, int[] visited, int cur) {
        if(visited[cur] == 1) return false;
        if(visited[cur] == 2) return true;

        visited[cur] = 1;
        for(int num : adj[cur]) {
            if(!dfs(adj, visited, num)) return false;
        }
        visited[cur] = 2;
        return true;
}
    public static boolean canFinish(int numCourses, int[][] prerequisites) {
        List<Integer>[] adj = new ArrayList[numCourses];
        for(int i=0;i<numCourses;i++) {
            adj[i] = new ArrayList<>();
        }

        for(int[] arr : prerequisites) {
            int c1 = arr[0];
            int c2 = arr[1];

            adj[c1].add(c2);
        }

        int[] visited = new int[numCourses];
        for(int i=0;i<numCourses;i++) {
            if(visited[i] == 0 && !dfs(adj, visited, i)) {
                return false;
            }
        }
        return true;
    }
    public static void main(String[] args) {
        int[][] arr = {{1,0}};
        System.out.println(canFinish(2, arr));
    }
}

File: 2025-09-25/Course_Schedule.py
from typing import List


class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        dict = {}

        for a, b in prerequisites:
            if a in dict.keys():
                dict[a].append(b)
            else:
                dict[a] = [b]

            if b not in dict.keys():
                dict[b] = []

        print(dict)
        def dfs(key, visited):
            if visited[key] == 1:
                return False

            if visited[key] == 2:
                return True

            visited[key] = 1
            for prerequisite in dict[key]:
                if not dfs(prerequisite, visited):
                    return False

            visited[key] = 2
            return True


        visited = [0] * numCourses
        for key in dict.keys():
            if not dfs(key, visited):
                return False

        print(True)
        return True

sol = Solution()

sol.canFinish(2, [[1,0]])
sol.canFinish(2, [[1,0],[0,1]])
sol.canFinish(5, [[1,4],[2,4],[3,1],[3,2]])